### PR TITLE
SINGA-200 - Implement Encoder and Decoder for data pre-processing

### PR DESCRIPTION
**Pull request by dbxinj**
Add base class Encoder and Decoder.
Add inheritance class Image2JPG_Encoder and Image2JPG_Decoder, which use opencv to encode and decode image source files.
One test problem is that encoded images is not exactly the same as the input image because of loss of information.

### CODE REVIEW

File: include/singa/io/encoder.h [hunk 1]
```diff
@@ -30,13 +31,15 @@ class Encoder {
   public:
     Encoder() { }
     virtual ~Encoder() { }
+    
+    virtual void Setup(const LayerConf& conf) = 0;
```
Comment (nudles) [inline include/singa/io/encoder.h]: why use LayerConf?

File: src/io/jpg2proto_encoder.cc [hunk 2]
```diff
@@ -16,21 +16,24 @@
...
+
+  std::string JPG2ProtoEncoder::Encode(vector<Tensor>& data) {
     // suppose data[0]: data, data[1]: label
     // data[0] has a shape as {height, width, channel}
     CHECK_EQ(data[0].nDim(), 3u);
     int height = data[0].shape()[0];
     int width = data[0].shape()[1];
     int channel = data[0].shape()[2];
-    Mat mat = Mat(height, width, CV_8UC3);
+    cv::Mat mat = cv::Mat(height, width, CV_8UC3, cv::Scalar(0, 0, 0));
```
Comment (nudles) [inline src/io/jpg2proto_encoder.cc]: Do you need cv2 or cv?

Comment (dbxinj) [inline src/io/jpg2proto_encoder.cc]: cv2
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/imgproc/imgproc.hpp>
namespace is still cv

File: include/singa/io/encoder.h (hunk 1 above)
Comment (dbxinj) [inline include/singa/io/encoder.h]: suppose some parameters are set in LayerConf
currently not sure about what params we will support for users to config

Comment (nudles) [inline include/singa/io/encoder.h]: do not make such assumption.
encoders and decoders should be independent from layer.

File: src/io/jpg2proto_encoder.cc (hunk 2 above)
Comment (nudles) [inline src/io/jpg2proto_encoder.cc]: then may need to check the link path. Is it opencv2 or opencv?

File: src/io/jpg2proto_encoder.cc [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    CHECK_EQ(data[0].nDim(), 3u);
+    int height = data[0].shape()[0];
+    int width = data[0].shape()[1];
+    int channel = data[0].shape()[2];
+    cv::Mat mat = cv::Mat(height, width, CV_8UC3, cv::Scalar(0, 0, 0));
+    CHECK_EQ(height, mat.rows);
+    CHECK_EQ(width, mat.cols);
+    CHECK_EQ(channel, mat.channels());
+
+    if (data[0].data_type() != kInt)
```
Comment (nudles) [inline src/io/jpg2proto_encoder.cc]: is this check necessary?

Comment (dbxinj) [inline src/io/jpg2proto_encoder.cc]: the first check is necessary, we want user to give a 3D input tensor for each raw image.
the rest three checks are not.

**Pull request merged**